[Muhammad Khair Bin Abdul Rashid] iP

muhammad-khair opened the pull request:
# Kayu PR

> JUST ANOTHER PR

## Present work done
- Rename mascot from ~~Duke~~ to **Kayu**
- Add commands _list_, _help_, _bye_, _delete_, _done_, _find_, _todo_, _event_, _deadline_
- Change ***greeting logo print***  

### Some spam here
1. Just do it
2. I'm loving it
3. LGTM

``` java
public static void main(String[] args) {
    Kayu kayu = new Kayu();
    kayu.run();
    System.exit(0);
}
```

## Self-checklist for GFMD task
- [x] heading
- [x] bullet list
- [x] numbered list
- [x] fenced code block with syntax highlighting
- [x] task list
- [x] emoji
- [x] blockquote
- [x] hyperlink
- [x] inline code
- [x] text formatting: bold, italics, strikethrough etc.

👁️👅👁️ 
> Link to full forked repo: [Click me](https://github.com/muhammad-khair/ip)

Commit:
Add Gradle support

muhammad-khair committed:
Refactor logo to static variable and method

muhammad-khair committed:
Add greet, run, echo and exit methods

muhammad-khair committed:
Add todo list and addToList method

muhammad-khair committed:
Add displayList method

muhammad-khair committed:
Add task class

muhammad-khair committed:
Integrate Task class to Duke

muhammad-khair committed:
Add markTaskAsDone method

muhammad-khair committed:
Task: rename fields, refactor getters and setters

muhammad-khair committed:
Task: refactor out status icon from toString to getStatusIcon method

muhammad-khair committed:
Duke: add simple comments for indexing

muhammad-khair committed:
Duke: rename action param for addToList method and add SCANNER.close in exit method

muhammad-khair committed:
Merge branch 'add-gradle-support'

muhammad-khair committed:
Add Todo class

muhammad-khair committed:
Add Deadline class

muhammad-khair committed:
Add Event class

muhammad-khair committed:
Deadline: change deadline field to by

muhammad-khair committed:
Remove overloaded constructor for Task and child classes

muhammad-khair committed:
Clean string formatting of Task and child classes

muhammad-khair committed:
Duke: rearrange methods in class and extend LINE_SPLIT field

muhammad-khair committed:
Duke: add checkUserInput method and change addToList to take in Task classes as input

muhammad-khair committed:
Duke: Change markTaskAsDone method to take in integer instead of string

muhammad-khair committed:
Duke: change message to echo in displayList method

muhammad-khair committed:
Add test input in input.txt and expected output in EXPECTED.TXT

muhammad-khair committed:
Revert "Add test input in input.txt and expected output in EXPECTED.TXT"

This reverts commit 9707e237398cd562866968b83408df0ebe8d2229.

muhammad-khair committed:
Task: change class to abstract class

muhammad-khair committed:
Duke: add message to displayList method

muhammad-khair committed:
Duke: refactor greeting

muhammad-khair committed:
Task: change status icon

muhammad-khair committed:
Refactor echo codes to Echoer class

muhammad-khair committed:
Duke: move exit method to main

muhammad-khair committed:
Task: add space to status icon

muhammad-khair committed:
Refactor task related methods from Duke to TaskManager

muhammad-khair committed:
Refactor task creation from Duke to TaskManager

muhammad-khair committed:
Add test input in input.txt and expected output in EXPECTED.TXT

muhammad-khair committed:
Refactor method names in echoer and add error template message

muhammad-khair committed:
Echoer: change error template

muhammad-khair committed:
Refactor Tasks into own package task

muhammad-khair committed:
Add TaskManagerException

muhammad-khair committed:
Refactor error logs to exceptions

muhammad-khair committed:
Change TaskManager to return Tasks instead of void

muhammad-khair committed:
Refactor printLogo into greet method

muhammad-khair committed:
Modify markTaskAsDone

muhammad-khair committed:
Added tests to check for error-prone inputs

muhammad-khair committed:
Add deleteTask functionality

muhammad-khair committed:
TaskManager: remove empty list exception for listTasks and change it to a string message

muhammad-khair committed:
Duke: Refactor greet and exit lines into main

muhammad-khair committed:
Refactor Echoer as DukeChatBot and make its methods non-static

muhammad-khair committed:
TaskManager: remove exception in listTasks' method signature

muhammad-khair committed:
Refactor key splitter into respecitve Tasks

muhammad-khair committed:
Edit splitter field for tasks

muhammad-khair committed:
TaskManager: add text to exception in splitUserInput method

muhammad-khair committed:
Text-ui-test: update tests

muhammad-khair committed:
Duke: refactor duke to have non-static methods

muhammad-khair committed:
Add javadocs

muhammad-khair committed:
Refactor DukeChatBot to ChatBot

muhammad-khair committed:
Refactor ChatBot and TaskManager to services package

muhammad-khair committed:
Refactor TaskManagerException to DukeException

muhammad-khair committed:
TaskManager: Edit javadocs

muhammad-khair committed:
Add enum Command

muhammad-khair committed:
Add Instruction class

muhammad-khair committed:
Add Parser service class

muhammad-khair committed:
Merge servives that use enum Command together

muhammad-khair committed:
Update test cases

muhammad-khair committed:
Change logo

muhammad-khair committed:
Add javadocs

muhammad-khair committed:
ChatBot: refactor strings to static fields and modify logo

muhammad-khair committed:
Clean Command parsing logic

muhammad-khair committed:
TaskManager: add assert statements and clean code

muhammad-khair committed:
Parser: clean logic and add more error strings

muhammad-khair committed:
Update expected test outcomes

muhammad-khair committed:
Remove redundant trim methods for user input

muhammad-khair committed:
TaskManager class: add initialiseList method

muhammad-khair committed:
TaskManager class: add loading task from file logic

muhammad-khair committed:
Add data folder in .gitignore

muhammad-khair committed:
Duke class: add loadTasks method call

muhammad-khair committed:
Modify tasks to have keyword for file parsing

muhammad-khair committed:
Add save task to file logic

muhammad-khair committed:
TaskManager class: add updateToFile method

muhammad-khair committed:
TaskManager class: add removeFromFile method

muhammad-khair committed:
Add DateTime recognition for Event and Deadline

muhammad-khair committed:
Merge branch 'branch-Level-7'

muhammad-khair committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/service/TaskManager.java
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java

muhammad-khair committed:
Rename TaskManager to TaskList

muhammad-khair committed:
Rename tasks toSavedString method to toEncodedString

muhammad-khair committed:
Add Storage class

muhammad-khair committed:
Clean Task classes

muhammad-khair committed:
Add DateTimeFormat utils class

muhammad-khair committed:
Refactor Parser class to own package

muhammad-khair committed:
Add StorageException class

muhammad-khair committed:
Refactor file logic from TaskList to Storage class

muhammad-khair committed:
Clean Task classes

muhammad-khair committed:
Refactor Commands to own package

muhammad-khair committed:
ChatBot class: change info to print for none-empty messages

muhammad-khair committed:
Change logic for Parser

muhammad-khair committed:
Migrate logic from Parser and TaskList to Commands

muhammad-khair committed:
Clean AddTaskCommand methods

muhammad-khair committed:
Clean code in Duke, ChatBot and Storage

muhammad-khair committed:
Rename commands package to command

muhammad-khair committed:
Text-UI-Tests: refactor to test directory

muhammad-khair committed:
Rename utils package to util

muhammad-khair committed:
Rename command package to commands

muhammad-khair committed:
TaskList class: change inequality sign for addTask method

muhammad-khair committed:
Refactor packages

muhammad-khair committed:
Add DeadlineTest class

muhammad-khair committed:
Add tests for Task classes

muhammad-khair committed:
Refactor DateTimeFormat to parser package

muhammad-khair committed:
Clean logic

muhammad-khair committed:
Add test classes and resources

muhammad-khair committed:
Rename program as Kayu and clean up

muhammad-khair committed:
ChatBot class: change error emoji symbol

muhammad-khair committed:
Change gradle configurations

muhammad-khair committed:
Gradle: add release version

muhammad-khair committed:
Add JavaDocs

muhammad-khair committed:
Clean code to follow standards

muhammad-khair committed:
Add FindCommand and associated logic

muhammad-khair committed:
FindCommand class: use static imports to reduce code length

muhammad-khair committed:
Refactor static variables to static imports when possible

muhammad-khair committed:
Update tests cases with FindCommand

muhammad-khair committed:
Merge branch 'A-JavaDoc'

muhammad-khair committed:
Merge branch 'A-CodingStandard'

muhammad-khair committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/test/java/kayu/parser/ParserTest.java

muhammad-khair committed:
Add more javadocs

muhammad-khair committed:
Add HelpCommand

muhammad-khair committed:
Remove unused import statement

muhammad-khair committed:
Rename methods to comply with code standards

muhammad-khair committed:
Modify build.gradle

muhammad-khair committed:
Add checkstyle.xml for gradle

muhammad-khair committed:
Clean javadocs

muhammad-khair committed:
Clean codestyle

muhammad-khair committed:
Modify trailing whitespace regex in checkstyle.xml

shermannws commented on src/main/java/kayu/commands/CommandMessage.java:
```diff
@@ -0,0 +1,31 @@
+public class CommandMessage {
+
+    // Message formats.
+    protected static final String MESSAGE_EMPTY_LIST = "List is empty, try adding some tasks first.";
```
I like how you adhered to the coding conventions here where constant names are uppercase and underscores are used to separate words.

shermannws commented on src/main/java/kayu/task/Event.java:
```diff
@@ -0,0 +1,101 @@
+ */
+public class Event extends Task {
+
+    /** Keyword for Event {@link kayu.task.Task} class. */
```
"Key Word" & "Keyword" have both been used in the codes/comments, perhaps you might want to consider standardising it?

muhammad-khair replied:
Resolved, went with "keyword" as the standard.

shermannws commented on src/main/java/kayu/storage/Storage.java:
```diff
@@ -0,0 +1,158 @@
+    protected List<Task> decodeAll(List<String> taskLines) throws StorageException {
+        List<Task> savedTaskList = new ArrayList<>();
+        
+        for (String stringTask : taskLines) {
```
You might want to consider standardising the syntax for this kind of **for loop** statements. Most of the other segments of your code do not have a whitespace before the colon, you might want to consider standardising to that format 😄

muhammad-khair replied:
Yeap good catch! Removed whitespace.

shermannws commented on src/main/java/kayu/commands/EmptyCommand.java:
```diff
@@ -0,0 +1,28 @@
+public class EmptyCommand extends Command {
+
+    /**
+     * Initializes a EmptyCommand- {@link kayu.commands.Command}.
```
Not a major issue, just a small grammatical error to use "an" instead of "a" before **EmptyCommand** in the Javadoc. Might want to ensure that the Javadoc is free of grammatical error. However, most of your Javadoc is already really well done, just a minor one here. 😄

muhammad-khair replied:
Resolved grammatical errors.

shermannws commented on src/main/java/kayu/commands/EventCommand.java:
```diff
@@ -0,0 +1,49 @@
+    public static final String COMMAND_WORD = "event";
+
+    /**
+     * Initializes a Event- {@link kayu.commands.AddTaskCommand}.
```
Similar Grammatical error here for "a" vs "an". Might want to check through your entire Javadoc.

shermannws reviewed (commented):
I think your Kayu code is really nicely organised with no obvious and huge coding violations, Good job on following the Java coding standard and I feel that there were a lot to learn from reviewing your code 👍 Keep up the good work, you're doing great!

muhammad-khair committed:
Initialise ui package with templates

muhammad-khair committed:
Route command outputs to ui

muhammad-khair committed:
Integrate Kayu to ui package and Launcher

muhammad-khair committed:
Clean code style

muhammad-khair committed:
Clean messages and task string formats

muhammad-khair commented:
Thank you so much @shermannws for reviewing! (especially when there's a lot of lines here compared to others haha) Will take note of the comments you mentioned and fix them in the following commits :)

muhammad-khair committed:
Merge branch 'branch-Level-10'

muhammad-khair committed:
Clean javadocs

muhammad-khair committed:
Clean codestyle

muhammad-khair committed:
Add varargs in FindCommand

muhammad-khair committed:
Refactor storage into command's execute method

muhammad-khair committed:
Replace 'key word' to 'keyword' in javadocs

muhammad-khair committed:
Change font of user input

yunpeng1234 commented on src/main/java/kayu/Kayu.java:
```diff
@@ -0,0 +1,108 @@
+    private final Storage storage = new Storage();
+    private final Logger logger = new Logger();
+    
+    private boolean isRecentCommandBye = false;
+
+    /**
+     * Returns whether the recent command parsed and executed is a {@link kayu.commands.ByeCommand}.
+     *
+     * @return Boolean true if recent command is a {@link kayu.commands.ByeCommand}, else false.
+     */
+    public boolean isRecentCommandBye() {
+        return isRecentCommandBye;
+    }
```
Maybe instead of making the isRecentCommandBye private, you can make the variable package protected and just call on it for other functions

yunpeng1234 commented on src/main/java/kayu/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,94 @@
+        for (DateTimeFormatter formatter: dateFormatterList) {
+            try {
+                return LocalDate.parse(dateString, formatter);
+            } catch (DateTimeParseException exception) {
+                // do nothing
+            }
```
Maybe can throw the Duke Exception in the catchblock to make it more clear to the reader that parsing LocalDate-> DukeException,

yunpeng1234 commented on src/main/java/kayu/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,94 @@
+        for (DateTimeFormatter formatter: timeFormatterList) {
+            try {
+                return LocalTime.parse(timeString, formatter);
+            } catch (DateTimeParseException exception) {
+                // do nothing
```
Same comment as the previous portion

yunpeng1234 commented on src/main/java/kayu/commands/Command.java:
```diff
@@ -0,0 +1,92 @@
+        return commandParams;
+    }
+
+    /**
+     * Checks if the Command instance is a {@link kayu.commands.ByeCommand}
+     * using its {@link #commandType}.
+     *
+     * @return Boolean true if is {@link kayu.commands.ByeCommand}, else false.
+     */
+    public boolean isBye() {
+        return commandType.equals(CommandType.BYE);
+    }
+
```
Could possibly override this method in the subclasses where Bye command class return true while all other command  classes return false.

yunpeng1234 commented on src/main/java/kayu/commands/CommandType.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Helps {@link kayu.commands.Command} by directing the {@link kayu.service.TaskList} to the right operations.
+ */
+public enum CommandType {
+    BYE,
+    DEADLINE,
+    DELETE,
+    DONE,
+    EMPTY,
+    EVENT,
+    FIND,
+    HELP,
+    INVALID,
+    LIST,
+    TODO;
+}
```
This class could be shifted into the Command class to make the command type variable more clear there.

yunpeng1234 commented on src/main/java/kayu/service/TaskList.java:
```diff
@@ -0,0 +1,145 @@
+    protected static final String FULL_CAPACITY_ERROR_MESSAGE = "Unable to execute as list is full.";
+    protected static final String EMPTY_LIST_ERROR_MESSAGE = "Unable to execute as list is empty.";
+
+    protected static final int MAX_STORAGE = 100;
```
Not sure if the max size is necessary as the point of the arrayList is so that it can grow

yunpeng1234 reviewed (commented):
Good code overall, there isn't any major issues with the code quality. LGTM

I would like to suggests that maybe only save the tasks when uses inputs a "bye" command.  This allows the code to be shorter but I do get your implementation of it as a way to quick save the code.

muhammad-khair committed:
Add assertions

muhammad-khair committed:
Refactor code into small readable methods

muhammad-khair committed:
Refactor Command logic to smaller methods

Current Command logic code blocks are not so readable. Some methods
have lengthy logic which makes it hard to comprehend.

Extracting code that can be generalised and placed in new method
signatures allows readability of code.

Doing such will allow future extensions to Commands to be implemented
with ease.

muhammad-khair committed:
Add assertion for Command creation

muhammad-khair committed:
Add enableAssertions field in build.gradle

muhammad-khair committed:
Merge pull request #1 from muhammad-khair/branch-A-Assertions

Add assertions

muhammad-khair committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/kayu/commands/AddTaskCommand.java
#	src/main/java/kayu/commands/Command.java

muhammad-khair committed:
Merge pull request #2 from muhammad-khair/branch-A-CodeQuality

Clean code quality

muhammad-khair committed:
Replace some for- loops with streams

muhammad-khair committed:
Add CI using .github/workflows/gradle.yml

muhammad-khair committed:
Code quality: perform refactoring across code base

There are inconsistencies with static variable message naming and
implementation of Commands.

Such inconsistencies can cause complications when working on the
code base further and can spiral to other inconsistencies if not
resolved.

Extracting code into smaller method chunks and standardising static
variable naming for messages allows us to comprehend the code
base better.

Let's,
* update relevant packages to follow a set standard
* improve the relationship and dependencies of Command classes

muhammad-khair committed:
D-Notes extension: add Note and NoteCommand logic

muhammad-khair committed:
D-Notes extension: add new Note Commands

muhammad-khair committed:
Update checkstyle whitespace regex

muhammad-khair committed:
Update header comments

muhammad-khair committed:
Update test code to comply with checkstyle

muhammad-khair committed:
Add Ui.png to docs

muhammad-khair committed:
Set theme jekyll-theme-minimal

muhammad-khair committed:
Update README.md

muhammad-khair committed:
Set theme jekyll-theme-slate

muhammad-khair committed:
Merge remote-tracking branch 'origin/master'

muhammad-khair committed:
Add user guide

muhammad-khair committed:
Update user guide

muhammad-khair committed:
Resolve table format in user guide

muhammad-khair committed:
Set theme jekyll-theme-midnight

muhammad-khair committed:
Update version

muhammad-khair committed:
Update docs README.md
